micro_epsilon_scancontrol_driver/include/micro_epsilon_scancontrol_driver/driver.h

```diff
@@ -91,7 +91,7 @@ namespace scancontrol_driver
 
 
             // Driver objects
-            CInterfaceLLT* device_interface;
+            std::shared_ptr<CInterfaceLLT> device_interface_ptr;
```

Comment:
Late, but: if there is no intention to actually *share* ownership of this pointer, and there can only be a single instance of it (ie: you cannot open the same device twice), perhaps using a `std::unique_ptr` would be more appropriate.

It would very clearly indicate that there can be only a single owner of the device handle.

Reply:
Did do that at first, but @IKapitaniuk suggested to actually use the `shared_ptr` instead ;)

Reply:
From scanning the code I don't see any place where the device handle is being passed as an argument, nor is it ever shared between two entities.

It's only ever used by dereferencing the pointer.

Seems like a good candidate for a `unique_ptr`.

Perhaps @IKapitaniuk can add his rationale here?

Reply:
Ok, regarding my rationale.

Of course, there is a guideline to prefer unique_ptr over shared_ptr, for instance [R.21](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-unique), because it is simpler and faster.

So, we don't care about the performance overhead, since we create the pointer only once.
Do we need to ensure the unique ownership, I don't think so, otherwise the only way is to use _const unique_ptr_, otherwise the pointer still can be stolen by std::move operator. 

My point is either we are strict and prohibit everything or make it as flexible as possible.
In this particular case, there is no much difference, but shared_ptr is just a more flexible solution for future extension.

Reply:
> Do we need to ensure the unique ownership, I don't think so, otherwise the only way is to use const unique_ptr, otherwise the pointer still can be stolen by std::move operator.

while this is certainly true (the last part), besides *enforcement*, we can also use types to *communicate* intent to (future) programmers (which will likely be us, but at a later time).

The use of `std::unique_ptr` would communicate the fact that this class is the one owning the device handle, that there can not be two owners of a single device handle at the same time, that things like copy-ctors must take this into account, and that we expect to see a certain pattern of usage when it comes to the device handle.

> My point is either we are strict and prohibit everything or make it as flexible as possible. In this particular case, there is no much difference, but shared_ptr is just a more flexible solution for future extension.

Flexibility is a good thing to have, but my main consideration would be: is there a context in which it would make sense for the device handle (an active one) to be shared among entities, to potentially be used at the same time? If not: a `shared_ptr` seems like it would not be the right thing to use here.

Reply:
> Flexibility is a good thing to have, but my main consideration would be: is there a context in which it would make sense for the device handle (an active one) to be shared among entities, to potentially be used at the same time? If not: a `shared_ptr` seems like it would not be the right thing to use here.

It's a very good point. Frankly speaking, I don't have an appropriate answer to this question. I would say the shared device handle might be of help if we going to use a separate thread(s) for configuring the device on the fly or for getting some statistic data. I think it is possible because Aravis supports multithreading, though not sure how good that support.

@dave992 Could you imagine any use case when you need running a few instances of ScanControlDriver attaching to the same hardware device? Or, in general, Is it possible to connect to the device if it is already connected to the driver, i.e. connect two separate nodes to the same device? If so, there is a context I am not aware of.

Reply:
Note: the question is not so much whether it would be technically possible to share the ownership of the handle.

It seems the question comes down to: is it, in the context of the class which owns the handle, ever possible or desirable for it to share the ownership of the handle it has with some other entity? And: would sharing make sense, or should another instance of the class be created?

Reply:
> @dave992 Could you imagine any use case when you need running a few instances of ScanControlDriver attaching to the same hardware device? Or, in general, Is it possible to connect to the device if it is already connected to the driver, i.e. connect two separate nodes to the same device? If so, there is a context I am not aware of.

No, I do not see any cases where multiple instances of `ScanControlDriver` connect to the same hardware device. I am not sure if the SDK even allows this (I do not see specific error codes for this), but even if it is possible I think it is undesirable to do so as the Interface is not notified of changes to important settings (i.e. the callback functions). 

> It seems the question comes down to: is it, in the context of the class which owns the handle, ever possible or desirable for it to share the ownership of the handle it has with some other entity? And: would sharing make sense, or should another instance of the class be created?

No, I don't think it makes sense to share the ownership of the handle with another entity as that somewhat defeats the purpose of the driver. The only case where I might see this happening is if a functionality is not implemented in the driver but is available via the SDK. However, this should in my opinion then be implemented in the Driver instead of retrieving the handle and doing it locally in your own program.

Reply:
At this point, I think it indeed is better to use the `unique_ptr`, even if only it is for communicating the intent of the use of the handle.

Reply:
Then I agree as well. `unique_ptr` is a better option.